### ISSUE TITLE
Why does LIME sample from synthetic data?

### ISSUE THREAD

**Issue by aeftimia** [marcotcr/lime#463]
# My Understanding

From the [original paper](https://arxiv.org/abs/1602.04938), it seems to objective of LIME is to train a simple model (such as a ridge regressor) with a loss function that places higher weight on data drawn closer to the reference sample. This makes a lot of sense to me, and reminds me of building a first order Taylor series approximation to the original model about the reference instance.

However, when I [peak under the hood](https://github.com/marcotcr/lime/blob/2ba75c188dcffe3e926c093efc5d03a0d51692b6/lime/lime_tabular.py#L517), it seems that by default, samples used to train this ridge regressor are drawn from a normal distribution with a diagonal covariance matrix that is fit to the raw data. It seems there is an option to shift the mean of this normal distribution to be centered on the sample of interest, but at that point your scale may no longer make sense.

If I understand correctly, you're training a ridge regressor on 

<img src="https://render.githubusercontent.com/render/math?math=\mathbb{E}_{x \sim \mathrm{synthetic}}\left(\mathrm{model}(x)-\mathrm{lime}(x)\right)^2 \mathrm{weight}_{\mathrm{ref}}(x)-\Omega\left(\mathrm{lime}\right)">

, where <img src="https://render.githubusercontent.com/render/math?math=\mathrm{model}"> is the original model, <img src="https://render.githubusercontent.com/render/math?math=\mathrm{lime}"> is by default a linear model, and <img src="https://render.githubusercontent.com/render/math?math=\Omega\left(\mathrm{lime}\right)"> is your ridge regularizer. Meanwhile, <img src="https://render.githubusercontent.com/render/math?math=\mathrm{synthetic}"> is your synthetic data distribution--namely your Gaussian with a diagonal covariance matrix (plus what appears to be multinomial distributions for categorical variables). <img src="https://render.githubusercontent.com/render/math?math=\mathrm{weight}_\mathrm{ref}"> is the weight a corresponding entry within the `sample_weight` array passed to, say, [Sklearn's Ridge regressor](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html). It would be derived from a distance function to a particular reference instance, <img src="https://render.githubusercontent.com/render/math?math=\mathrm{ref}">.

# My Question

If you're already weighing your loss function using some distance function, why draw synthetic samples in the first place? Why not just train on the original dataset?

<img src="https://render.githubusercontent.com/render/math?math=\mathbb{E}_{x \sim \mathrm{raw data}}\left(\mathrm{model}(x)-\mathrm{lime}(x)\right)^2 \mathrm{weight}_\mathrm{ref}(x)-\Omega\left(\mathrm{lime}\right)">

I apologize if I'm missing something very obvious. However, understanding the motivation for the synthetic data would greatly help me interpret the output of LIME in practice.

**Comment by marcotcr** [comment]
We don't particularly care about the behavior of the model on the training data, as most modern models will have 0 training error (in which case we are training a linear model on the original training data, weighted by how close they are to the example being explained).

We could potentially use the validation set, but we are assuming that will not be large enough, or diverse enough to have enough weight near the example we are trying to explain.

There is a tradeoff between being as realistic as possible to the original data distribution and surfacing behaviors about what we actually care about. One of the uses of explanation is surfacing bad behaviors caused by biases in the original dataset, which is one more reason to sample from out-of-distribution data (but not too far out :) )

**Comment by aeftimia** [comment]
Thank you very much for your explanation @marcotcr! I asked my original question in part because my dataset would not be very well approximated by diagonal Gaussians.

> We could potentially use the validation set, but we are assuming that will not be large enough, or diverse enough to have enough weight near the example we are trying to explain.

If the validation set were large and diverse enough to assess the accuracy of the model, is there any reason to suspect it wouldn't be large or diverse enough to train LIME?

> One of the uses of explanation is surfacing bad behaviors caused by biases in the original dataset, which is one more reason to sample from out-of-distribution data (but not too far out :) )

This makes sense. Though is there cause for concern that a sufficiently mismatched synthetic distribution would create new biases? For example, if my data wasn't particularly concentrated near its mean, I would imagine training a new model on synthetic data drawn from a gaussian fit would overrepresent examples near the mean during the training process.

I could imagine a procedure in which the weights are adjusted to partially counter this bias.

<img src="https://render.githubusercontent.com/render/math?math=\mathbb{E}_{x \sim \mathrm{synthetic}}\left(\mathrm{model}(x)-\mathrm{lime}(x)\right)^2 \mathrm{weight}_\mathrm{ref}(x)\frac{\hat{p}_\mathrm{rawdata}(x)}{p_\mathrm{synthetic}(x)}-\Omega\left(\mathrm{lime}\right)">

, where <img src="https://render.githubusercontent.com/render/math?math=\hat{p}_\mathrm{rawdata}(x)"> might be some better approximation to the true density of the raw data (say from KNN). But by the law of large numbers, this would be equivalent to sampling from <img src="https://render.githubusercontent.com/render/math?math=\hat{p}_\mathrm{rawdata}(x)"> directly.

**Comment by marcotcr** [comment]
> If the validation set were large and diverse enough to assess the accuracy of the model, is there any reason to suspect it wouldn't be large or diverse enough to train LIME?

I think so. Consider the very simple example of explaining an outliner in an 1d dataset. You may have a large enough dataset to check average accuracy, but not enough samples near the particular example you want to explain. I think this reasoning holds if we generalize this intuition to real data.

> This makes sense. Though is there cause for concern that a sufficiently mismatched synthetic distribution would create new biases?

Yes, for sure. That's why it's hard to set these defaults - there are these trade-offs : ). The hope is that these biases are mitigated by the kernel function more so than the alternative.
